File: Sample.java
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class MyQueue {
    private Stack<Integer> inStack;
    private Stack<Integer> outStack;

    public MyQueue() {
        this.inStack = new Stack<>();
        this.outStack = new Stack<>();
    }
    
    public void push(int x) {
        inStack.push(x);
    }
    
    public int pop() {
        peek();
        return outStack.pop();
    }
    
    public int peek() {
        if(outStack.isEmpty()) {
            while(!inStack.isEmpty()) {
                outStack.push(inStack.pop());
            }
        }
        return outStack.peek();
    }
    
    public boolean empty() {
        return inStack.isEmpty() && outStack.isEmpty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */

// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class MyHashMap {

    class Node {
        int key;
        int value;
        Node next;

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    Node[] storage;

    public MyHashMap() {
        this.storage = new Node[10000];
    }

    private int hash(int key) {
        return key%10000;
    }

    private Node search(Node head, int key) {
        Node prev = null;
        Node curr = head;

        while (curr!=null && curr.key!=key) {
            prev = curr;
            curr= curr.next;
        }

        return prev;
    }
    
    public void put(int key, int value) {
        int idx = hash(key);

        if(storage[idx] == null) {
            storage[idx] = new Node(-1,-1);
        }

        Node prev = search(storage[idx], key);

        if(prev.next != null) {
            prev.next.value = value;
        } 
        else {
            prev.next = new Node(key, value);
        }
    }
    
    public int get(int key) {
        int idx = hash(key);

        if(storage[idx] == null) {
            return -1;
        }

        Node prev = search(storage[idx], key);

        if(prev.next == null) return -1;
        return prev.next.value;
    }
    
    public void remove(int key) {
        int idx = hash(key);

        if(storage[idx] == null) {
            return;
        }

        Node prev = search(storage[idx], key);
        if(prev.next == null) return;
        Node temp = prev.next;
        prev.next = prev.next.next;
        temp.next = null;
    }
}

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */

File: Sample.py
// Time Complexity :
// Space Complexity :
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :
// Your code here along with comments explaining your approach

# Design-2

Explain your approach in **three sentences only** at top of your code


## Problem 1: (https://leetcode.com/problems/implement-queue-using-stacks/)
# My approach
Initialize two stacks 
Keep on push the element to the push_stack 
When pop comes push all the items to pop stack  from push_stack and pop out the Last element.
If its empty return False
class MyQueue:

    def __init__(self):
        self.push_stack = []
        self.pop_stack =[]

    def push(self, x: int) -> None:
        self.push_stack.append(x)

    def pop(self) -> int:
        if self.empty(): return
        if len(self.pop_stack):
            return self.pop_stack.pop()
        else:
            while len(self.push_stack):
                self.pop_stack.append(self.push_stack.pop())
        return self.pop_stack.pop()

    def peek(self) -> int:
        if self.empty(): return
        if len(self.pop_stack):
            return self.pop_stack[-1]
        else:
            while len(self.push_stack):
                self.pop_stack.append(self.push_stack.pop())
        return self.pop_stack[-1]

    def empty(self) -> bool:
        return len(self.push_stack)==False and len(self.pop_stack) == False
        
Time Complexity : 0(1), push: 0(1), Pop - 0(n), peek - 0(n), empty - 0(1), Space Complexity: 0(n)
Did this code successfully run on Leetcode : yes
Any problem you faced while coding this : nope

## Problem 2:
Design Hashmap (https://leetcode.com/problems/design-hashmap/)
# My approach
Intialize the data
put updates the value associated with that key.
get key returns the value to which the specified key is mapped
removes the mapping of the specified value key if this map contains a mapping for the key
if no mapping for the key is -1.

class MyHashMap:
    
    def __init__(self):
        self.my_array = [-1 for i in range(1000001)]
        
    def put(self, key: int, value: int) -> None:
        self.my_array[key] = value
        
    def get(self, key: int) -> int:
        return self.my_array[key]
        
    def remove(self, key: int) -> None:
        self.my_array[key] = -1

Time Complexity : 0(1), Space Complexity: 0(n)
Did this code successfully run on Leetcode : yes
Any problem you faced while coding this : nope



